src/libraries/System.Linq/src/System/Linq/OrderedEnumerable.cs

```diff
@@ -359,10 +359,43 @@ internal override int CompareAnyKeys(int index1, int index2)
             return (_descending != (c > 0)) ? 1 : -1;
         }
 
+        private int CompareAnyKeys_DefaultComparer_NoNext_Ascending(int index1, int index2)
+        {
+            Debug.Assert(_comparer == Comparer<TKey>.Default);
+            Debug.Assert(_next is null);
+            Debug.Assert(!_descending);
+
+            TKey[]? keys = _keys;
+            Debug.Assert(keys != null);
+
+            int c = Comparer<TKey>.Default.Compare(keys[index1], keys[index2]);
+            return
+                c == 0 ? index1 - index2 : // ensure stability of sort
+                c;
+        }
+
+        private int CompareAnyKeys_DefaultComparer_NoNext_Descending(int index1, int index2)
+        {
+            Debug.Assert(_comparer == Comparer<TKey>.Default);
+            Debug.Assert(_next is null);
+            Debug.Assert(_descending);
+
+            TKey[]? keys = _keys;
+            Debug.Assert(keys != null);
+
+            int c = Comparer<TKey>.Default.Compare(keys[index2], keys[index1]);
+            return
+                c == 0 ? index1 - index2 : // ensure stability of sort
+                c;
+        }
+
         private int CompareKeys(int index1, int index2) => index1 == index2 ? 0 : CompareAnyKeys(index1, index2);
 
         protected override void QuickSort(int[] keys, int lo, int hi) =>
-            new Span<int>(keys, lo, hi - lo + 1).Sort(CompareAnyKeys);
+            new Span<int>(keys, lo, hi - lo + 1).Sort(
+                !typeof(TKey).IsValueType || _next is not null || _comparer != Comparer<TKey>.Default ? new Comparison<int>(CompareAnyKeys) :
+                _descending ? new Comparison<int>(CompareAnyKeys_DefaultComparer_NoNext_Descending) :
+                new Comparison<int>(CompareAnyKeys_DefaultComparer_NoNext_Ascending));
```

Comment:
I take it that the QuickSort method is only called once in the lifetime of this object and we're fine with allocating a new delegate every time?

Reply:
It is. "fine" is a strong term :-) We'd prefer not to, but we don't have a better option right now. Note this was allocating a delegate already; I just changed the syntax to be explicit about it.